docs/2019-04-23-Applicative-Kore.md

```diff
@@ -1,222 +1,286 @@
+
 ### Patterns
 
 ```
 <atom-pattern> ::=
-| <id>        // can be either <element-variable> or <symbol>, decided by verifier
-| <sharp-id>  // <set-variable>
-| "\bottom"
-| "(" <pattern> ")"
+| <id>                // can mean either an element variable, 
+                      // or a constant, or a notation (alias);
+                      // this has to be decided later by the *verifier*
+| <sharp-id>          // a set variable
+| "\bottom"           // the bottom
+| "(" <pattern> ")"   // parentheses used for grouping
 
 <app-pattern> ::=
 | <atom-pattern>
-| <app-pattern> <atom-pattern> // i.e., application is right associative
+| <app-pattern> <atom-pattern> // a b c is parsed as ((a b) c)
 
 <pattern> ::=
-| "\implies" <atom-pattern> <atom-pattern>
-| "\forall" <element-variable> <atom-pattern>
-| "\mu" <set-variable> <atom-pattern>
+| "\implies" <atom-pattern> <atom-pattern>     // implication
+| "\forall" <element-variable> <atom-pattern>  // FOL binders
+| "\mu" <set-variable> <atom-pattern>          // fixpoint binders
+
+<element-variable> ::= <id>
+<set-variable>     ::= <sharp-id>
 ```
 
 Some examples:
 
 ```
-\implies a        ... ... not wellformed; not enough argument
+\implies a        ... ... not wellformed: too few arguments
 \implies a b      ... ... wellformed
-\implies a b c    ... ... not wellformed; too many arguments
-\implies (a b) c  ... ... wellformed; (a b) is an application
+\implies a b c    ... ... not wellformed: too many arguments
+\implies (a b) c  ... ... wellformed
+```
+
+### Sentences and modules
+
+An AppKore module consisting of a sequence of _sentences_. _Import sentence_ imports a module; _symbol sentence_ declares a constant symbol; _notation sentence_ declares a notation (a.k.a. alias); _axiom sentence_ specifies a pattern as an axiom of the module; a _hooked symbol sentence_ declares a constant symbol whose semantics is not given by the axioms, but given by the implementation; a _hooked notation sentence_ declares a notation whose definition is not given in the declaration, but given by the implementation. 
+
+Each sentence/module is associated with zero, one, or more _attributes_. Attributes provide information and hints for the backends but cannot carry _additional_ semantic meaning. 
+
+```
+<sentence> ::=
+| <import-sentence>
+| <symbol-sentence>
+| <notation-sentence>
+| <axiom-sentence>
+| <hooked-symbol-sentence>
+| <hooked-notation-sentence>
+
+<import-sentence> ::= "import" <id> <attribute>
+<symbol-sentence> ::= "symbol" <id> <attribute>
+<notation-sentence> ::= "notation" <id> <ids> "=" <pattern> <attribute>
+<axiom-sentence> ::= "axiom" <pattern> <attribute>
+<hooked-symbol-sentence> ::= "hooked-symbol" <id> <attribute>
```

Comment:
Is it necessary to distinguish `hooked-symbol` from `symbol`? Backends may provide builtin support for any sorts and symbols, and the declaration syntax is not distinct as with `notation` and `hooked-notation`.